cmd/dbc/auth.go

```diff
@@ -0,0 +1,261 @@
+// Copyright 2025 Columnar Technologies Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/charmbracelet/bubbles/spinner"
+	tea "github.com/charmbracelet/bubbletea"
+	"github.com/cli/oauth/device"
+	"github.com/columnar-tech/dbc/auth"
+)
+
+type AuthCmd struct {
+	Login  *LoginCmd  `arg:"subcommand" help:"Authenticate with a driver index"`
+	Logout *LogoutCmd `arg:"subcommand" help:"Logout from a driver index"`
+}
+
+type LoginCmd struct {
+	URI      string `arg:"positional" help:"URL of the driver index to authenticate with"`
+	ClientID string `arg:"env:OAUTH_CLIENT_ID" help:"OAuth Client ID (can also be set via DBC_OAUTH_CLIENT_ID)"`
+	Audience string `arg:"env:OAUTH_AUDIENCE" help:"OAuth Audience (can also be set via DBC_OAUTH_AUDIENCE)"`
+	ApiKey   string `arg:"--api-key" help:"Authenticate using an API key instead of OAuth"`
+}
+
+func (l LoginCmd) GetModelCustom(baseModel baseModel) tea.Model {
+	if l.URI == "" {
+		l.URI = auth.DefaultOauthURI
+	}
+
+	if l.URI == auth.DefaultOauthURI {
+		if l.ClientID == "" {
+			l.ClientID = auth.DefaultOauthClientID
+		}
+		if l.Audience == "" {
+			l.Audience = auth.DefaultOauthAudience
+		}
+	}
+
+	s := spinner.New()
+	s.Spinner = spinner.MiniDot
+	return loginModel{
+		spinner:       s,
+		inputURI:      l.URI,
+		oauthClientID: l.ClientID,
+		oauthAudience: l.Audience,
+		apiKey:        l.ApiKey,
+		baseModel:     baseModel,
+	}
+}
+
+func (l LoginCmd) GetModel() tea.Model {
+	return l.GetModelCustom(
+		baseModel{
+			getDriverRegistry: getDriverRegistry,
+			downloadPkg:       downloadPkg,
+		},
+	)
+}
+
+type loginModel struct {
+	baseModel
+
+	spinner spinner.Model
+
+	inputURI      string
+	oauthClientID string
+	oauthAudience string
+	apiKey        string
+	tokenURI      *url.URL
+	parsedURI     *url.URL
+}
+
+func (m loginModel) Init() tea.Cmd {
+	if !strings.HasPrefix(m.inputURI, "https://") {
+		m.inputURI = "https://" + m.inputURI
+	}
+
+	u, err := url.Parse(m.inputURI)
+	if err != nil {
+		return errCmd("invalid URI provided: %w", err)
+	}
+
+	return tea.Batch(m.spinner.Tick, func() tea.Msg {
+		return u
+	})
+}
+
+func (m loginModel) authConfig() tea.Cmd {
+	return func() tea.Msg {
+		cfg, err := auth.GetOpenIDConfig(m.parsedURI)
+		if err != nil {
+			return fmt.Errorf("failed to get OpenID configuration: %w", err)
```

Comment:
The error we produce here feels slightly leaky:

```
$ go run ./cmd/dbc auth login
Error: failed to get OpenID configuration: Get "https://dbc-cdn-private.columnar.tech/.well-known/openid-configuration": dial tcp: lookup dbc-cdn-private.columnar.tech: no such host
exit status 1
```

The user doesn't need to know any implementation details of our OAuth flow, not that the URL or the path are secrets and not easily guessable. Thoughts? There's a good chance I'm being needlessly paranoid.

Reply:
follow up sounds fine. I'm not even 100% sure if we need to quiet this specific error.